client/my-sites/media-library/list-item.tsx

```diff
@@ -1,53 +1,59 @@
-/** @format */
-
 /**
  * External dependencies
  */
-
-import { assign, isEqual, noop, omit } from 'lodash';
-import PropTypes from 'prop-types';
+import { isEqual, noop } from 'lodash';
 import React from 'react';
 import classNames from 'classnames';
 
 /**
  * Internal dependencies
  */
 import Spinner from 'components/spinner';
-import Gridicon from 'gridicons';
+import Gridicon from 'components/gridicon';
 import ListItemImage from './list-item-image';
 import ListItemVideo from './list-item-video';
 import ListItemAudio from './list-item-audio';
 import ListItemDocument from './list-item-document';
 import { getMimePrefix } from 'lib/media/utils';
 import EditorMediaModalGalleryHelp from 'post-editor/media-modal/gallery-help';
+
 /**
  * Style dependencies
  */
 import './list-item.scss';
 
-export default class extends React.Component {
-	static displayName = 'MediaLibraryListItem';
-
-	static propTypes = {
-		media: PropTypes.object,
-		scale: PropTypes.number.isRequired,
-		maxImageWidth: PropTypes.number,
-		thumbnailType: PropTypes.string,
-		showGalleryHelp: PropTypes.bool,
-		selectedIndex: PropTypes.number,
-		onToggle: PropTypes.func,
-		onEditItem: PropTypes.func,
-		style: PropTypes.object,
-	};
+// TODO: move to lib/media/utils once it gets typed.
+interface MediaObject {
+	transient?: boolean;
+	file?: string;
+	[propName: string]: any;
+}
+type Media = string | MediaObject;
+// END TODO
+
+interface Props {
+	media?: Media;
+	scale: number;
+	maxImageWidth?: number;
+	thumbnailType?: string;
+	showGalleryHelp?: boolean;
+	selectedIndex?: number;
+	onToggle?: ( media: Media | undefined, shiftKey: boolean ) => any;
```

Comment:
> media: Media | undefined

Interesting. In Flow you can have "maybe" types e.g. `media: ?Media` and that would be equivalent to `media: Media | null | undefined` - but I guess not in TS. 🤔

Reply:
You can do that, with a caveat:

> Any optional parameters must follow required parameters.

https://www.typescriptlang.org/docs/handbook/functions.html#optional-and-default-parameters

```ts
function f( x?: string, y: string ) {} // TypeScript error - A required parameter cannot follow an optional parameter.

function g( x: string, y?: string ) {} // 👍 Optional after required.
```

Reply:
Yup, although Flow's a little more semantic/expressive in this case (or more generally). There's both a concept of "optional params" and "maybe types" e.g.

Optional (as in TS): `{ foo?: number }`
Maybe: `{ foo: ?number }`

So in the latter, `{}` would be invalid, but `{ foo: null }` or `{ foo: undefined }` would pass.

https://flow.org/en/docs/types/functions/#toc-optional-parameters
https://flow.org/en/docs/types/maybe/

Reply:
I don't think that's available in TS, is it, @sirreal ?

I do like the fact that this is explicit in TS, though, as it makes it easy to distinguish between similar, but different situations: `foo | null`, `foo | undefined` and `foo | null | undefined` are all different.

Reply:
> I do like the fact that this is explicit in TS, though, as it makes it easy to distinguish between similar, but different situations: foo | null, foo | undefined and foo | null | undefined are all different.

Yes, agree (yes & no actually). I think the dichotomy/difference is narrowing down (but that might just be wishful thinking). 😬

Not sure I'd do justice explaining any of the reasoning. Might just be convenience/idiomatic to Flow. I'd normally refer to @beaucollins for insight, but he's in a different realm now. :/

Reply:
I think the way the arguments are typed is good for TypeScript 👍 

Minor: Now that I'm looking closely, did you consider `void` return? Might be preferable.

```suggestion
	onToggle?: ( media: Media | undefined, shiftKey: boolean ) => void;
```

Reply:
> Minor: Now that I'm looking closely, did you consider `void` return? Might be preferable.

I generally tend to avoid being over-restrictive. This component doesn't *need* the return value to be `void`, as it simply ignores it, so it seems fine to me to type the return as `any`.

Is there a particular reason why we'd want to enforce `void`? Arrow functions make it very easy to "accidentally" return a value, and I'm not sure it's in our interest to flag that as an error. For example, imagine the following method inside of a consumer class:

```js
onToggle = ( media, shiftKey ) => this.media = media;
```

This "accidentally" returns a value and would have to be rewritten in a more verbose fashion to avoid doing so:

```js
onToggle = ( media, shiftKey ) => {
  this.media = media;
}
```

Reply:
> onToggle = ( media, shiftKey ) => this.media = media;

That's a good example. Not sure if it's a good idea though to develop types to accommodate accidental behaviours. 🤔 They should depict the actual expected functionality right i.e. that `onToggle` doesn't return anything in this case - while "any" is more akin to "something". I think `any` sounds more like code smell in a type system. 😬

Reply:
I guess it fundamentally boils down to how we see `any` as a return value to a function. I see it as "I don't care what your function returns", whereas from what I gather you see it as "I don't care what your function returns, as long as it returns something", right?

I wonder if there's a more explicit way of saying "I don't care what your function returns, even if it returns nothing"... 🤔

Reply:
I guess we could do `any | void`, but TypeScript simplifies that to `any`.

Reply:
> > Minor: Now that I'm looking closely, did you consider `void` return? Might be preferable.
> 
> I generally tend to avoid being over-restrictive. This component doesn't _need_ the return value to be `void`, as it simply ignores it, so it seems fine to me to type the return as `any`.

What's happening here is a bit beyond me and I don't fully understand why this is allowed. It's a minor point I don't think it's worth getting hung up on.

[They do recommend `void` as the return type for callbacks if you don't use the return value.](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#return-types-of-callbacks)

The takeaway is that `void` return on the callback doesn't restrict the callbacks that can be passed in, just on what you can do with the value returned from the callback.

For example:

```ts
const f = ( cb: (() => void)) => {
    const cbValue: void = cb();
    // can't do much with cbValue
}

// `callback: () => string` satisfies `() => void`??
const callback = (): string => 'a string value';
f( callback )

// No type errors in anything above
// Yet, none of these work 😕
const n: null = 1;      // 💥
const u: undefined = 1; // 💥
const v: void = 1;      // 💥
```

Why we can pass something of type `() => string` to a function that accepts type `() => void` isn't clear to me, but I think we should prefer `void` here.

Reply:
What @sirreal said: function that returns something is compatible with a `void` return type and can be passed as, e.g., event handler. It just means that the return value is not used. I did a similar thing recently when typing another event handler.

Reply:
> Why we can pass something of type () => string to a function that accepts type () => void isn't clear to me

A type theorist would call this a [covariant method return type](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)#Covariant_method_return_type), and could also talk about "contravariant method parameter types" 🙂 

If there is a type hierarchy where `Cat extends Animal`, a "function that returns `Animal`" type constraint can be type-safely satisfied with a function that returns `Cat`. All `Cat`s that are returned are always also `Animal`s.

Similarly, a "function that takes a `Cat` argument" type constraint can be satisfied with a function that takes an `Animal` argument. All `Cat`s that are passed are always also `Animal`s.

The direction of the arrow is different in each case and that's what the fancy "covariant" and "contravariant" words mean.

Reply:
Thank you both for your really extensive and detailed explanations! 🙇

So this was all a case of me completely misunderstanding what `void` means as a return value in this context, sorry 😅 I thought it meant that only functions without a return type are allowed, but I guess that's not the case!

If this is the case, it completely makes sense to switch the signature. Done!